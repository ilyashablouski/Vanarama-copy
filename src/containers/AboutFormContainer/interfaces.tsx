import { CreateUpdatePersonMutation as Mutation } from '../../../generated/CreateUpdatePersonMutation';

export interface IProps {
  onCompleted: (data: Mutation) => void;
  onLogInClick?: () => void;
  onRegistrationClick?: () => void;
  personUuid?: string;
  personLoggedIn?: boolean;
}
