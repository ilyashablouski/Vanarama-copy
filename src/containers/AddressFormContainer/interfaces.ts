export interface IAddressFormContainerProps {
  personId: string;
  onCompleted: () => void;
}
