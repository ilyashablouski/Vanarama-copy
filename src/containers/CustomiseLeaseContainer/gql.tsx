import { useQuery, gql, useLazyQuery, ApolloError } from '@apollo/client';
import {
  GetQuoteDetails,
  GetQuoteDetailsVariables,
} from '../../../generated/GetQuoteDetails';
import { IQuoteDataInputs } from './interfaces';

export const GET_QUOTE_DATA = gql`
  query GetQuoteDetails(
    $capId: ID!
    $vehicleType: VehicleTypeEnum
    $leaseType: LeaseTypeEnum
    $mileage: Int
    $trim: Int
    $colour: Int
    $term: Int
    $upfront: Int
  ) {
    quoteByCapId(
      capId: $capId
      mileage: $mileage
      term: $term
      upfront: $upfront
      vehicleType: $vehicleType
      leaseType: $leaseType
      colour: $colour
      trim: $trim
    ) {
      term
      funderId
      mileage
      upfront
      trim
      colour
      leadTime
      stock
      vehicleType
      leaseType
      processingFee
      nextBestPrice {
        maintained
        nonMaintained
      }
      leaseCost {
        monthlyRental
        initialRental
        excessMileage
      }
      maintenanceCost {
        monthlyRental
        initialRental
        excessMileage
      }
    }
  }
`;

export function useQuoteData({
  capId,
  vehicleType,
  mileage,
  term,
  upfront,
  leaseType,
  trim,
  colour,
}: IQuoteDataInputs) {
  return useQuery<GetQuoteDetails, GetQuoteDetailsVariables>(GET_QUOTE_DATA, {
    variables: {
      capId,
      vehicleType,
      mileage,
      term,
      upfront,
      leaseType,
      trim,
      colour,
    },
  });
}

export function useQuoteDataLazyQuery(
  onCompleted?: (data: GetQuoteDetails) => void,
  onError?: (error: ApolloError) => void,
) {
  return useLazyQuery<GetQuoteDetails, GetQuoteDetailsVariables>(
    GET_QUOTE_DATA,
    {
      onCompleted,
      onError,
    },
  );
}
