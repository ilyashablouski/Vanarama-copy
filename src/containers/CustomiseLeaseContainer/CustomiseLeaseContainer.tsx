import dynamic from 'next/dynamic';
import React, { useEffect, useState } from 'react';
import Modal from 'core/molecules/modal';
import CustomiseLease from '../../components/CustomiseLease/CustomiseLease';
import { useQuoteDataLazyQuery } from './gql';
import {
  LeaseTypeEnum,
  OpportunityTypeEnum,
  VehicleTypeEnum,
} from '../../../generated/globalTypes';
import { IProps } from './interfaces';
import { GetQuoteDetails } from '../../../generated/GetQuoteDetails';
import GoldrushFormContainer from '../GoldrushFormContainer';
import useFirstRenderEffect from '../../hooks/useFirstRenderEffect';
import Skeleton from '../../components/Skeleton';
import getLineItem from '../../utils/getLineItem';
import { useMobileViewport } from '../../hooks/useMediaQuery';
import { useTrim } from '../../gql/carpage';
import { Nullable, Nullish } from '../../types/common';
import { IOptionsList } from '../../types/detailsPage';
import {
  checkIsHotOffer,
  isFactoryOrderSelect,
  isHotOfferSelect,
} from '../../utils/helpers';

const Loading = dynamic(() => import('core/atoms/loading'), {
  loading: () => <Skeleton count={1} />,
});

const parseQuoteParams = (param?: string | null) =>
  parseInt(param || '', 10) || null;

const createEmptyQuoteData = (
  term: number | null,
  mileage: number | null,
  upfront: number | null,
  vehicleType: VehicleTypeEnum | null,
  trim: number | null,
  colour: number | null,
  leaseType: LeaseTypeEnum | null,
) => ({
  quoteByCapId: {
    term,
    mileage,
    upfront,
    vehicleType,
    trim: trim?.toString() || null,
    colour: colour?.toString() || null,
    leaseType,
    funderId: null,
    leadTime: null,
    stock: null,
    freeInsurance: null,
    stockBatchId: null,
    processingFee: null,
    nextBestPrice: null,
    maintenanceCost: null,
    leaseCost: {
      monthlyRental: null,
      initialRental: null,
      excessMileage: null,
    },
  },
});

const START_SCREEN_HEIGHT = 350;

// eslint-disable-next-line no-empty-pattern
const CustomiseLeaseContainer: React.FC<IProps> = ({
  quote,
  capId,
  vehicleType,
  derivativeInfo,
  leaseAdjustParams,
  leaseType,
  setLeaseType,
  setLeadTime,
  onCompleted,
  onCompletedCallBack,
  setLeaseScannerData,
  isPlayingLeaseAnimation,
  setIsPlayingLeaseAnimation,
  trimData,
  colourData,
  mileage,
  colour,
  setColour,
  setMileage,
  pickups,
  isShowFreeInsuranceMerch,
  isShowFreeHomeChargerMerch,
  roadsideAssistance,
  warrantyDetails,
  dataUiTestId,
}) => {
  const [isFactoryOrder, setIsFactoryOrder] = useState<boolean | undefined>(
    isFactoryOrderSelect(colourData, `${colour}`),
  );
  const [isHotOffer, setIsHotOffer] = useState<Nullish<boolean>>(
    isHotOfferSelect(colourData, `${colour}`),
  );
  const [quoteData, setQuoteData] = useState<
    GetQuoteDetails | null | undefined
  >(quote || null);
  const [upfront, setUpfront] = useState<number | null>(
    quote?.quoteByCapId?.upfront || null,
  );
  const [term, setTerm] = useState<number | null>(
    quote?.quoteByCapId?.term || null,
  );
  const [trim, setTrim] = useState<number | null>(
    parseQuoteParams(quote?.quoteByCapId?.trim),
  );
  const [trimList, setTrimList] = useState<Nullable<IOptionsList[]>>(
    checkIsHotOffer(trimData, isHotOffer, isFactoryOrder),
  );

  const [maintenance, setMaintenance] = useState<boolean | null>(null);
  const [isModalShowing, setIsModalShowing] = useState<boolean>(false);
  const [isInitPayModalShowing, setIsInitPayModalShowing] = useState<boolean>(
    false,
  );
  const [isRestoreLeaseSettings, setIsRestoreLeaseSettings] = useState(false);
  const [isInitialLoading, setIsInitialLoading] = useState<boolean>(false);
  const [showCallBackForm, setShowCallBackForm] = useState<boolean>(false);
  const [isStartScreen, setIsStartScreen] = useState<boolean>(true);
  const isMobile = useMobileViewport();
  const [getTrim] = useTrim(result => {
    setTrimList(
      checkIsHotOffer(result.trimGroupList, isHotOffer, isFactoryOrder),
    );
  });
  const [getQuoteData, { loading }] = useQuoteDataLazyQuery(
    updatedQuote => {
      setIsRestoreLeaseSettings(false);
      setQuoteData(updatedQuote);
      if (updatedQuote.quoteByCapId?.trim) {
        setTrim(Number(updatedQuote.quoteByCapId?.trim));
      }
    },
    () =>
      setQuoteData(
        createEmptyQuoteData(
          term,
          mileage,
          upfront,
          vehicleType,
          trim,
          colour,
          leaseType,
        ),
      ),
  );

  useEffect(() => {
    let timerId: NodeJS.Timeout;

    if (isInitialLoading && isPlayingLeaseAnimation && !loading) {
      timerId = setTimeout(() => {
        setIsPlayingLeaseAnimation(false);
      }, 1000);
    }

    return () => {
      clearTimeout(timerId);
    };
  }, [
    isInitialLoading,
    isPlayingLeaseAnimation,
    setIsPlayingLeaseAnimation,
    loading,
  ]);

  useEffect(() => {
    setLeadTime(quoteData?.quoteByCapId?.leadTime || '');
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [quoteData]);

  const scrollChange = () => {
    if (isMobile) {
      setIsStartScreen(() => window.pageYOffset < START_SCREEN_HEIGHT);
    }
  };

  useEffect(() => {
    if (isMobile) {
      window?.addEventListener('scroll', scrollChange);
    }
    return () => window?.removeEventListener('scroll', scrollChange);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isMobile]);

  const currentQuoteTrim = quoteData?.quoteByCapId?.trim;
  const currentQuoteColour = quoteData?.quoteByCapId?.colour;

  useFirstRenderEffect(() => {
    getQuoteData({
      variables: {
        capId: `${capId}`,
        vehicleType,
        mileage,
        term,
        upfront,
        leaseType,
        trim: trim || parseQuoteParams(currentQuoteTrim),
        colour: colour || parseQuoteParams(currentQuoteColour),
      },
    });

    if (!isRestoreLeaseSettings) {
      setIsPlayingLeaseAnimation(true);
    }
  }, [
    leaseType,
    upfront,
    mileage,
    term,
    trim,
    getQuoteData,
    vehicleType,
    capId,
  ]);

  useFirstRenderEffect(() => {
    getQuoteData({
      variables: {
        capId: `${capId}`,
        vehicleType,
        mileage,
        term,
        upfront,
        leaseType,
        colour: colour || parseQuoteParams(currentQuoteColour),
      },
    });

    getTrim({
      variables: {
        capId: `${capId}`,
        vehicleType,
        colourId: colour as number,
      },
    });

    if (!isRestoreLeaseSettings) {
      setIsPlayingLeaseAnimation(true);
    }
  }, [colour]);

  const lineItemData = getLineItem({
    vehicleTypeValue: vehicleType,
    maintenanceValue: maintenance,
    mileageValue: mileage,
    upfrontValue: upfront,
    trimValue: trim,
    termValue: term,
    capId,
    quoteData,
    colourData,
    trimData: trimList,
  });

  useEffect(() => {
    if (setLeaseScannerData) {
      setLeaseScannerData({
        maintenance,
        quoteByCapId: quoteData?.quoteByCapId,
        isDisabled: isPlayingLeaseAnimation,
        stateVAT: leaseType === LeaseTypeEnum.PERSONAL ? 'inc' : 'exc',
        endAnimation: () => {
          setIsInitialLoading(true);
        },
        requestCallBack: () => {
          setShowCallBackForm(true);
        },
        onSubmit: () => {
          onCompleted({
            leaseType: leaseType.toUpperCase() as LeaseTypeEnum,
            lineItems: [lineItemData],
          });
        },
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    quoteData,
    leaseType,
    getQuoteData,
    isPlayingLeaseAnimation,
    maintenance,
  ]);

  if (!quoteData) {
    return (
      <div
        className="pdp--sidebar"
        style={{ minHeight: '40rem', display: 'flex', alignItems: 'center' }}
      >
        <Loading size="xlarge" />
      </div>
    );
  }

  const terms = leaseAdjustParams?.terms.map(String);
  const upfronts = leaseAdjustParams?.upfronts.map(String);
  const leaseTypes = [LeaseTypeEnum.PERSONAL, LeaseTypeEnum.BUSINESS];

  const defaultTermValue = quote?.quoteByCapId?.term ?? null;
  const defaultUpfrontValue = quote?.quoteByCapId?.upfront ?? null;
  const defaultMileageValue = quote?.quoteByCapId?.mileage ?? null;

  // - show POA form in case if during first render (SSR) monthlyRental is not returned
  // - show POA form in case if monthlyRental returned but colors or trim lists are empty
  // - all other cases of errors from server will be handled in LeaseScanner
  if (
    !quote?.quoteByCapId?.leaseCost?.monthlyRental ||
    colourData?.length === 0 ||
    trimList?.length === 0
  ) {
    // eslint-disable-next-line no-console
    console.error({
      colourData,
      trimList,
      monthlyRental: quoteData?.quoteByCapId?.leaseCost?.monthlyRental,
    });

    return (
      <GoldrushFormContainer
        dataUiTestId={`${dataUiTestId}_goldrush-form`}
        termsAndConditions
        isPostcodeVisible={vehicleType !== VehicleTypeEnum.CAR}
        capId={capId}
        opportunityType={OpportunityTypeEnum.QUOTE}
        vehicleType={vehicleType}
      />
    );
  }

  return (
    <>
      <CustomiseLease
        term={term}
        terms={terms ?? []}
        upfront={upfront}
        upfronts={upfronts ?? []}
        defaultTermValue={defaultTermValue}
        defaultUpfrontValue={defaultUpfrontValue}
        defaultMileageValue={defaultMileageValue}
        leaseType={leaseType}
        leaseTypes={leaseTypes}
        mileages={leaseAdjustParams?.mileages ?? []}
        setLeaseType={setLeaseType}
        setMileage={setMileage}
        setUpfront={setUpfront}
        setColour={setColour}
        setTerm={setTerm}
        setTrim={setTrim}
        data={quoteData}
        capId={capId}
        mileage={mileage || quoteData?.quoteByCapId?.mileage}
        trim={String(trim)}
        derivativeInfo={derivativeInfo}
        colour={String(colour)}
        trimData={trimList}
        colourData={colourData}
        isPlayingLeaseAnimation={isPlayingLeaseAnimation}
        setIsPlayingLeaseAnimation={setIsPlayingLeaseAnimation}
        leaseAdjustParams={leaseAdjustParams}
        maintenance={maintenance}
        setMaintenance={setMaintenance}
        isModalShowing={isModalShowing}
        setIsModalShowing={setIsModalShowing}
        isInitPayModalShowing={isInitPayModalShowing}
        setIsInitPayModalShowing={setIsInitPayModalShowing}
        setIsInitialLoading={setIsInitialLoading}
        setIsRestoreLeaseSettings={setIsRestoreLeaseSettings}
        lineItem={lineItemData}
        isStartScreen={isStartScreen}
        onSubmit={values => onCompleted(values)}
        showCallBackForm={() => setShowCallBackForm(true)}
        pickups={pickups}
        isShowFreeInsuranceMerch={isShowFreeInsuranceMerch}
        isShowFreeHomeChargerMerch={isShowFreeHomeChargerMerch}
        roadsideAssistance={roadsideAssistance}
        warrantyDetails={warrantyDetails}
        setIsHotOffer={setIsHotOffer}
        setIsFactoryOrder={setIsFactoryOrder}
      />
      <Modal
        className="-mt-000 callBack"
        show={showCallBackForm}
        onRequestClose={() => setShowCallBackForm(false)}
      >
        <GoldrushFormContainer
          dataUiTestId={`${dataUiTestId}_goldrush-form`}
          isPostcodeVisible={vehicleType !== VehicleTypeEnum.CAR}
          capId={capId}
          isCallBackForm
          opportunityType={OpportunityTypeEnum.QUOTE}
          vehicleType={vehicleType}
          onCompleted={() => {
            onCompletedCallBack();
            setShowCallBackForm(false);
          }}
        />
      </Modal>
    </>
  );
};

export default CustomiseLeaseContainer;
