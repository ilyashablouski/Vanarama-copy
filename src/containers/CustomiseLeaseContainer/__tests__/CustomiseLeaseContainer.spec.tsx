import React from 'react';
import preloadAll from 'jest-next-dynamic';
import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CustomiseLeaseContainer from '../CustomiseLeaseContainer';
import { useQuoteDataLazyQuery } from '../gql';
import { IProps } from '../interfaces';
import { VehicleTypeEnum } from '../../../../generated/globalTypes';
import { useOpportunityCreation } from '../../GoldrushFormContainer/gql';

jest.mock('../gql');
jest.mock('../../GoldrushFormContainer/gql');

const getComponent = (props: IProps) => {
  return renderer.create(<CustomiseLeaseContainer {...props} />).toJSON();
};

describe('<CustomiseLeaseContainer />', () => {
  beforeEach(async () => {
    await preloadAll();
  });
  (useOpportunityCreation as jest.Mock).mockReturnValue([
    () => {},
    { loading: false },
  ]);

  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  it('should show data correctly', async () => {
    const data = {
      quoteByCapId: {
        colour: '13990',
        leadTime: '14-21 Day Delivery',
        leaseType: 'PERSONAL',
        maintained: {
          monthlyRental: 61.75,
          initialRental: 61.75,
          excessMileage: 0,
        },
        mileage: 12000,
        nonMaintained: {
          monthlyRental: 605.95,
          initialRental: 605.95,
          excessMileage: 14.76,
        },
        leaseCost: {
          monthlyRental: 123,
        },
        processingFee: 0,
        stock: 'Brand New - ',
        term: 24,
        trim: '112981',
        upfront: 1,
        vehicleType: 'CAR',
      },
    } as any;
    (useQuoteDataLazyQuery as jest.Mock).mockReturnValue([
      () => {},
      {
        loading: false,
        data,
        error: undefined,
        refetch: jest.fn(),
        leaseType: 'Personal',
        setLeaseType: jest.fn(),
      },
    ]);

    render(
      <CustomiseLeaseContainer
        quote={data}
        capId={84429}
        colourData={[{ optionId: 13990, label: 'Solid - Polar white' }]}
        trimData={[{ optionId: 104562, label: 'Leather - Cranberry red' }]}
        vehicleType={VehicleTypeEnum.CAR}
        onCompletedCallBack={jest.fn()}
        setLeadTime={jest.fn()}
        isDisabled={false}
        setIsDisabled={jest.fn()}
        leaseAdjustParams={{
          mileages: [6000, 8000, 10000, 12000, 15000, 20000, 25000, 30000],
          terms: [24, 36, 48, 60],
          upfronts: [1, 3, 6, 9, 12],
        }}
        derivativeInfo={{
          name: '',
          manufacturer: {
            name: '',
            slug: '',
          },
          model: {
            name: '',
            slug: '',
          },
          range: {
            name: '',
            slug: '',
          },
          transmission: {
            name: 'Manual',
          },
          bodyStyle: {
            name: 'Hatchback',
          },
          bodyType: {
            name: 'Hatchback',
            slug: 'hatchback',
          },
          fuelType: {
            name: 'Diesel',
          },
          colours: [{ id: '13990', optionDescription: 'Solid - Polar white' }],
          technicals: [
            {
              categoryDescription: 'Weight and Capacities',
              derivativeId: '84429',
              effectiveFrom: '2019-07-01T00:00:00.000Z',
              effectiveTo: null,
              id: '3',
              technicalDescription: 'Minimum Kerbweight',
              technicalLongDescription: 'Minimum Kerbweight',
              value: '1515',
              unit: 'kg',
            },
          ],
          standardEquipments: [
            {
              categoryDescription: 'Body Glass',
              derivativeId: '84429',
              effectiveFrom: '2018-06-01T00:00:00.000Z',
              effectiveTo: null,
              genericDescription: null,
              id: '1475',
              optionDescription: 'Green tinted glass',
              optionLongDescription: 'Green tinted glass',
            },
          ],
          trims: [
            { id: '104562', optionDescription: 'Leather - Cranberry red' },
          ],
        }}
        leaseType="Business"
        setLeaseType={jest.fn()}
        onCompleted={jest.fn()}
        mileage={6000}
        setMileage={jest.fn()}
      />,
    );

    expect(screen.getByText('12000 Miles'));

    expect(screen.getByText('24 Months - 2 Years'));

    fireEvent.click(screen.getByText('Business'));

    await waitFor(() => {
      expect(screen.getByText('12000 Miles'));
    });

    await waitFor(() => {
      expect(screen.getByText('24 Months - 2 Years'));
    });

    await waitFor(() => {
      expect(screen.getByText('PM exc. VAT'));
    });
  });

  it('should show data correctly with data', async () => {
    const data = {
      quoteByCapId: {
        colour: '13990',
        leadTime: '14-21 Day Delivery',
        leaseType: 'BUSINESS',
        maintained: {
          monthlyRental: 61.75,
          initialRental: 61.75,
          excessMileage: 0,
        },
        mileage: 12000,
        nonMaintained: {
          monthlyRental: 605.95,
          initialRental: 605.95,
          excessMileage: 14.76,
        },
        processingFee: 0,
        stock: 'Brand New - ',
        term: 24,
        trim: '112981',
        upfront: 1,
        vehicleType: 'CAR',
      },
    } as any;
    (useQuoteDataLazyQuery as jest.Mock).mockReturnValue([
      () => {},
      {
        loading: false,
        data,
        error: undefined,
        refetch: jest.fn(),
      },
    ]);

    const tree = getComponent({
      quote: data,
      capId: 84429,
      colourData: [{ optionId: 13990, label: 'Solid - Polar white' }],
      trimData: [{ optionId: 104562, label: 'Leather - Cranberry red' }],
      vehicleType: VehicleTypeEnum.CAR,
      setLeadTime: jest.fn(),
      isDisabled: false,
      setIsDisabled: jest.fn(),
      leaseAdjustParams: {
        mileages: [6000, 8000, 10000, 12000, 15000, 20000, 25000, 30000],
        terms: [24, 36, 48, 60],
        upfronts: [1, 3, 6, 9, 12],
      },
      derivativeInfo: {
        name: '',
        manufacturer: {
          name: '',
          slug: '',
        },
        model: {
          name: '',
          slug: '',
        },
        range: {
          name: '',
          slug: '',
        },
        transmission: {
          name: 'Manual',
        },
        bodyStyle: {
          name: 'Hatchback',
        },
        bodyType: {
          name: 'Hatchback',
          slug: 'hatchback',
        },
        fuelType: {
          name: 'Diesel',
        },
        colours: [{ id: '13990', optionDescription: 'Solid - Polar white' }],
        technicals: [
          {
            categoryDescription: 'Weight and Capacities',
            derivativeId: '84429',
            effectiveFrom: '2019-07-01T00:00:00.000Z',
            effectiveTo: null,
            id: '3',
            technicalDescription: 'Minimum Kerbweight',
            technicalLongDescription: 'Minimum Kerbweight',
            value: '1515',
            unit: 'kg',
          },
        ],
        standardEquipments: [
          {
            categoryDescription: 'Body Glass',
            derivativeId: '84429',
            effectiveFrom: '2018-06-01T00:00:00.000Z',
            effectiveTo: null,
            genericDescription: null,
            id: '1475',
            optionDescription: 'Green tinted glass',
            optionLongDescription: 'Green tinted glass',
          },
        ],
        trims: [{ id: '104562', optionDescription: 'Leather - Cranberry red' }],
      },
      leaseType: 'Personal',
      setLeaseType: jest.fn(),
      onCompleted: jest.fn(),
      onCompletedCallBack: jest.fn(),
      mileage: 6000,
      setMileage: jest.fn(),
    });
    expect(tree).toMatchSnapshot();
  });

  it('should show data correctly with data', async () => {
    (useQuoteDataLazyQuery as jest.Mock).mockReturnValue([
      () => {},
      {
        loading: true,
        data: undefined,
        error: undefined,
        refetch: jest.fn(),
      },
    ]);

    const tree = getComponent({
      capId: 84429,
      vehicleType: VehicleTypeEnum.CAR,
      colourData: [{ optionId: 13990, label: 'Solid - Polar white' }],
      trimData: [{ optionId: 104562, label: 'Leather - Cranberry red' }],
      isDisabled: false,
      setIsDisabled: jest.fn(),
      onCompletedCallBack: jest.fn(),
      setLeadTime: jest.fn(),
      leaseAdjustParams: {
        mileages: [6000, 8000, 10000, 12000, 15000, 20000, 25000, 30000],
        terms: [24, 36, 48, 60],
        upfronts: [1, 3, 6, 9, 12],
      },
      derivativeInfo: {
        name: '',
        model: {
          name: '',
          slug: '',
        },
        manufacturer: {
          name: '',
          slug: '',
        },
        range: {
          name: '',
          slug: '',
        },
        transmission: {
          name: 'Manual',
        },
        bodyStyle: {
          name: 'Hatchback',
        },
        bodyType: {
          name: 'Hatchback',
          slug: 'hatchback',
        },
        fuelType: {
          name: 'Diesel',
        },
        colours: [{ id: '13990', optionDescription: 'Solid - Polar white' }],
        technicals: [
          {
            categoryDescription: 'Weight and Capacities',
            derivativeId: '84429',
            effectiveFrom: '2019-07-01T00:00:00.000Z',
            effectiveTo: null,
            id: '3',
            technicalDescription: 'Minimum Kerbweight',
            technicalLongDescription: 'Minimum Kerbweight',
            value: '1515',
            unit: 'kg',
          },
        ],
        standardEquipments: [
          {
            categoryDescription: 'Body Glass',
            derivativeId: '84429',
            effectiveFrom: '2018-06-01T00:00:00.000Z',
            effectiveTo: null,
            genericDescription: null,
            id: '1475',
            optionDescription: 'Green tinted glass',
            optionLongDescription: 'Green tinted glass',
          },
        ],
        trims: [{ id: '104562', optionDescription: 'Leather - Cranberry red' }],
      },
      leaseType: 'Personal',
      setLeaseType: jest.fn(),
      onCompleted: jest.fn(),
      mileage: 6000,
      setMileage: jest.fn(),
    });
    expect(tree).toMatchSnapshot();
  });

  it('should show data correctly', async () => {
    const data = {
      quoteByCapId: {
        colour: '13990',
        leadTime: '14-21 Day Delivery',
        leaseType: 'PERSONAL',
        maintained: {
          monthlyRental: 61.75,
          initialRental: 61.75,
          excessMileage: 0,
        },
        mileage: 12000,
        nonMaintained: {
          monthlyRental: 605.95,
          initialRental: 605.95,
          excessMileage: 14.76,
        },
        leaseCost: {
          monthlyRental: 123,
        },
        processingFee: 0,
        stock: 'Brand New - ',
        term: 24,
        trim: '112981',
        upfront: 1,
        vehicleType: 'CAR',
      },
    } as any;
    (useQuoteDataLazyQuery as jest.Mock).mockReturnValue([
      () => {},
      {
        loading: false,
        data,
        onCompleted: () => data,
        error: undefined,
        refetch: jest.fn(),
        leaseType: 'Personal',
        setLeaseType: jest.fn(),
      },
    ]);

    render(
      <CustomiseLeaseContainer
        quote={data}
        capId={84429}
        colourData={[{ optionId: 13990, label: 'Solid - Polar white' }]}
        trimData={[{ optionId: 104562, label: 'Leather - Cranberry red' }]}
        onCompletedCallBack={jest.fn()}
        vehicleType={VehicleTypeEnum.CAR}
        setLeadTime={jest.fn()}
        isDisabled={false}
        setIsDisabled={jest.fn()}
        leaseAdjustParams={{
          mileages: [6000, 8000, 10000, 12000, 15000, 20000, 25000, 30000],
          terms: [24, 36, 48, 60],
          upfronts: [1, 3, 6, 9, 12],
        }}
        derivativeInfo={{
          name: '',
          manufacturer: {
            name: '',
            slug: '',
          },
          model: {
            name: '',
            slug: '',
          },
          range: {
            name: '',
            slug: '',
          },
          transmission: {
            name: 'Manual',
          },
          bodyStyle: {
            name: 'Hatchback',
          },
          bodyType: {
            name: 'Hatchback',
            slug: 'hatchback',
          },
          fuelType: {
            name: 'Diesel',
          },
          colours: [{ id: '13990', optionDescription: 'Solid - Polar white' }],
          technicals: [
            {
              categoryDescription: 'Weight and Capacities',
              derivativeId: '84429',
              effectiveFrom: '2019-07-01T00:00:00.000Z',
              effectiveTo: null,
              id: '3',
              technicalDescription: 'Minimum Kerbweight',
              technicalLongDescription: 'Minimum Kerbweight',
              value: '1515',
              unit: 'kg',
            },
          ],
          standardEquipments: [
            {
              categoryDescription: 'Body Glass',
              derivativeId: '84429',
              effectiveFrom: '2018-06-01T00:00:00.000Z',
              effectiveTo: null,
              genericDescription: null,
              id: '1475',
              optionDescription: 'Green tinted glass',
              optionLongDescription: 'Green tinted glass',
            },
          ],
          trims: [
            { id: '104562', optionDescription: 'Leather - Cranberry red' },
          ],
        }}
        leaseType="Business"
        setLeaseType={jest.fn()}
        onCompleted={jest.fn()}
        mileage={6000}
        setMileage={jest.fn()}
      />,
    );

    fireEvent.click(screen.getByText('Request a Call Back'));

    await waitFor(() => {
      expect(screen.getByText('Please Fill In Your Details'));
    });
  });
});
