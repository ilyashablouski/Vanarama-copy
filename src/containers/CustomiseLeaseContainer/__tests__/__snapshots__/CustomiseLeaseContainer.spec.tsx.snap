// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomiseLeaseContainer /> should show data correctly with data 1`] = `
<div
  className="pdp--sidebar"
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div>
      <span
        className="heading -black -small"
      >
        FACTORY ORDER
      </span>
      <span
        className="text -small -darker"
      >
        Availability: 12 Weeks (Avg)
      </span>
    </div>
    <div>
      <div
        className="price -xlarge -black"
      >
        <div
          className="price--inner"
        >
          <span
            className="price--sub"
          >
            £
          </span>
          <span
            className="price--pounds"
          >
            POA
          </span>
        </div>
      </div>
    </div>
  </div>
  <form
    className="form"
    data-testid="goldrush-form"
    onSubmit={[Function]}
  >
    <span
      className="heading -black -lead"
    >
      Get Your Quote Now
    </span>
    <span
      className="text -mb-400 -regular -darker"
    >
      We’ll be in touch within 1-2 business hours
    </span>
    <div
      className="formgroup"
    >
      <label
        className="formgroup--legend"
        htmlFor="goldrush-form_full-name"
      >
        Full Name
      </label>
      <div
        className="formgroup--content"
      >
        <div
          className="textinput regular"
        >
          <input
            className="textinput--native regular"
            data-testid="goldrush-form_full-name"
            id="goldrush-form_full-name"
            name="fullName"
            placeholder=""
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="formgroup"
    >
      <label
        className="formgroup--legend"
        htmlFor="goldrush-form_email"
      >
        Email Address
      </label>
      <div
        className="formgroup--content"
      >
        <div
          className="textinput regular"
        >
          <input
            className="textinput--native regular"
            data-testid="goldrush-form_email"
            id="goldrush-form_email"
            name="email"
            placeholder=""
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="formgroup"
    >
      <label
        className="formgroup--legend"
        htmlFor="goldrush-form_phone-number"
      >
        Phone Number
      </label>
      <div
        className="formgroup--content"
      >
        <div
          className="textinput regular"
        >
          <input
            className="textinput--native regular"
            data-testid="goldrush-form_phone-number"
            id="goldrush-form_phone-number"
            name="phoneNumber"
            placeholder=""
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="formgroup"
    >
      <label
        className="formgroup--legend"
      >
        Please Confirm
      </label>
      <div
        className="formgroup--content"
      >
        <div
          className="checkbox -teal"
        >
          <input
            className="checkbox--native visually-hidden"
            data-testid="aboutConsent"
            id="consent"
            name="consent"
            type="checkbox"
          />
          <label
            className="checkbox--label"
            htmlFor="consent"
          >
            <span
              className="text -regular -darker"
            >
              Keep me updated on the latest deals & offers
            </span>
          </label>
        </div>
        <div
          className="checkbox -teal"
        >
          <input
            className="checkbox--native visually-hidden"
            data-testid="aboutTermsAndCons"
            id="termsAndCons"
            name="termsAndCons"
            type="checkbox"
          />
          <label
            className="checkbox--label"
            htmlFor="termsAndCons"
          >
            <span
              className="text -regular -darker"
            >
              I agree to the 
              <a
                className="link -teal"
                href="/legal/terms-and-conditions"
                target="_blank"
              >
                Terms and Conditions
              </a>
            </span>
          </label>
        </div>
      </div>
    </div>
    <p
      className="text -xsmall -darker"
    >
      Vanarama collects the contact information you provide to us to contact you about our products and services. You may unsubscribe from these communications at any time. For information on how to unsubscribe, as well as our privacy practices and commitment to protecting your privacy, please check out our Privacy Policy.
    </p>
    <button
      className="button -teal -lead -solid"
      data-testid="goldrush-form_submit"
      disabled={false}
      type="submit"
    >
      <div
        className="button--inner"
      >
        Get Quote Now
      </div>
    </button>
  </form>
  <div
    className="pdp--sidebar-promise"
  >
    <span
      className="text -regular -black"
    >
      We’ll beat any lease quote or we'll give you £100. 
    </span>
    <a
      className="link -success -small -plain"
      href="#"
    >
      Terms and Conditions apply.
    </a>
  </div>
</div>
`;

exports[`<CustomiseLeaseContainer /> should show data correctly with data 2`] = `
<div
  className="pdp--sidebar"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "minHeight": "40rem",
    }
  }
>
  <div
    className="loading -xlarge"
  >
    <svg
      className="loading--svg"
      viewBox="-75 -75 150 150"
    >
      <circle
        className="loading--stroke"
        cx="0"
        cy="0"
        r="37.5"
      />
    </svg>
  </div>
</div>
`;
